<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">@import url(https://themes.googleusercontent.com/fonts/css?kit=XGMkxXUZTA64h2imyzu79g);.lst-kix_e63skeaub1i7-3>li:before{content:"\0025cf   "}.lst-kix_e63skeaub1i7-4>li:before{content:"\0025cb   "}.lst-kix_e63skeaub1i7-1>li:before{content:"\0025cb   "}.lst-kix_e63skeaub1i7-2>li:before{content:"\0025a0   "}.lst-kix_e63skeaub1i7-5>li:before{content:"\0025a0   "}.lst-kix_e63skeaub1i7-6>li:before{content:"\0025cf   "}.lst-kix_e63skeaub1i7-0>li:before{content:"\0025cf   "}.lst-kix_e63skeaub1i7-7>li:before{content:"\0025cb   "}.lst-kix_e63skeaub1i7-8>li:before{content:"\0025a0   "}.lst-kix_i97szj75x3di-6>li{counter-increment:lst-ctn-kix_i97szj75x3di-6}ol.lst-kix_khhpt4pbx05-1.start{counter-reset:lst-ctn-kix_khhpt4pbx05-1 0}ul.lst-kix_2u1s0aut9bz5-1{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-0{list-style-type:none}.lst-kix_khhpt4pbx05-5>li{counter-increment:lst-ctn-kix_khhpt4pbx05-5}ul.lst-kix_2u1s0aut9bz5-3{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-2{list-style-type:none}ul.lst-kix_e63skeaub1i7-8{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-5{list-style-type:none}ul.lst-kix_e63skeaub1i7-7{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-4{list-style-type:none}ul.lst-kix_e63skeaub1i7-6{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-7{list-style-type:none}ul.lst-kix_e63skeaub1i7-5{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-6{list-style-type:none}ul.lst-kix_2u1s0aut9bz5-8{list-style-type:none}ol.lst-kix_khhpt4pbx05-7.start{counter-reset:lst-ctn-kix_khhpt4pbx05-7 0}.lst-kix_i97szj75x3di-8>li{counter-increment:lst-ctn-kix_i97szj75x3di-8}ol.lst-kix_i97szj75x3di-3.start{counter-reset:lst-ctn-kix_i97szj75x3di-3 0}.lst-kix_z7kpv721mdo1-8>li:before{content:"\0025a0   "}.lst-kix_khhpt4pbx05-1>li{counter-increment:lst-ctn-kix_khhpt4pbx05-1}ol.lst-kix_i97szj75x3di-2.start{counter-reset:lst-ctn-kix_i97szj75x3di-2 0}.lst-kix_z7kpv721mdo1-4>li:before{content:"\0025cb   "}.lst-kix_khhpt4pbx05-7>li{counter-increment:lst-ctn-kix_khhpt4pbx05-7}.lst-kix_z7kpv721mdo1-3>li:before{content:"\0025cf   "}ol.lst-kix_khhpt4pbx05-6.start{counter-reset:lst-ctn-kix_khhpt4pbx05-6 0}.lst-kix_i97szj75x3di-4>li{counter-increment:lst-ctn-kix_i97szj75x3di-4}.lst-kix_z7kpv721mdo1-5>li:before{content:"\0025a0   "}.lst-kix_z7kpv721mdo1-6>li:before{content:"\0025cf   "}ol.lst-kix_khhpt4pbx05-0.start{counter-reset:lst-ctn-kix_khhpt4pbx05-0 0}.lst-kix_z7kpv721mdo1-7>li:before{content:"\0025cb   "}.lst-kix_2u1s0aut9bz5-6>li:before{content:"\0025cf   "}.lst-kix_2u1s0aut9bz5-5>li:before{content:"\0025a0   "}.lst-kix_2u1s0aut9bz5-7>li:before{content:"\0025cb   "}.lst-kix_2u1s0aut9bz5-4>li:before{content:"\0025cb   "}.lst-kix_2u1s0aut9bz5-8>li:before{content:"\0025a0   "}.lst-kix_2u1s0aut9bz5-2>li:before{content:"\0025a0   "}.lst-kix_2u1s0aut9bz5-1>li:before{content:"\0025cb   "}.lst-kix_2u1s0aut9bz5-3>li:before{content:"\0025cf   "}ol.lst-kix_khhpt4pbx05-8{list-style-type:none}ol.lst-kix_khhpt4pbx05-7{list-style-type:none}ol.lst-kix_khhpt4pbx05-6{list-style-type:none}ol.lst-kix_khhpt4pbx05-5{list-style-type:none}ol.lst-kix_khhpt4pbx05-4{list-style-type:none}ol.lst-kix_khhpt4pbx05-3{list-style-type:none}ol.lst-kix_khhpt4pbx05-2{list-style-type:none}ol.lst-kix_khhpt4pbx05-1{list-style-type:none}ol.lst-kix_khhpt4pbx05-0{list-style-type:none}ol.lst-kix_i97szj75x3di-8.start{counter-reset:lst-ctn-kix_i97szj75x3di-8 0}ol.lst-kix_i97szj75x3di-7.start{counter-reset:lst-ctn-kix_i97szj75x3di-7 0}.lst-kix_kfnp1fjph7ti-2>li:before{content:"\0025a0   "}.lst-kix_kfnp1fjph7ti-6>li:before{content:"\0025cf   "}.lst-kix_i97szj75x3di-0>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-0,decimal) ". "}.lst-kix_jfoyp4fe4c9-1>li:before{content:"\0025cb   "}.lst-kix_jfoyp4fe4c9-5>li:before{content:"\0025a0   "}.lst-kix_kfnp1fjph7ti-0>li:before{content:"\0025cf   "}.lst-kix_kfnp1fjph7ti-8>li:before{content:"\0025a0   "}.lst-kix_i97szj75x3di-6>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-6,decimal) ". "}.lst-kix_jfoyp4fe4c9-7>li:before{content:"\0025cb   "}ol.lst-kix_khhpt4pbx05-5.start{counter-reset:lst-ctn-kix_khhpt4pbx05-5 0}.lst-kix_csthsd1cfrto-3>li:before{content:"\0025cf   "}.lst-kix_rysblip21ldc-1>li:before{content:"\0025cb   "}ol.lst-kix_i97szj75x3di-4.start{counter-reset:lst-ctn-kix_i97szj75x3di-4 0}.lst-kix_ova5lqw9xm1i-1>li:before{content:"\0025cb   "}.lst-kix_2u1s0aut9bz5-0>li:before{content:"\0025cf   "}.lst-kix_csthsd1cfrto-5>li:before{content:"\0025a0   "}.lst-kix_i97szj75x3di-4>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-4,lower-latin) ". "}.lst-kix_ova5lqw9xm1i-3>li:before{content:"\0025cf   "}.lst-kix_csthsd1cfrto-7>li:before{content:"\0025cb   "}.lst-kix_rysblip21ldc-3>li:before{content:"\0025cf   "}.lst-kix_rysblip21ldc-5>li:before{content:"\0025a0   "}.lst-kix_kfnp1fjph7ti-4>li:before{content:"\0025cb   "}.lst-kix_i97szj75x3di-2>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-2,lower-roman) ". "}.lst-kix_jfoyp4fe4c9-3>li:before{content:"\0025cf   "}.lst-kix_ova5lqw9xm1i-7>li:before{content:"\0025cb   "}ul.lst-kix_jfoyp4fe4c9-7{list-style-type:none}ul.lst-kix_jfoyp4fe4c9-6{list-style-type:none}ul.lst-kix_jfoyp4fe4c9-5{list-style-type:none}.lst-kix_i97szj75x3di-2>li{counter-increment:lst-ctn-kix_i97szj75x3di-2}ul.lst-kix_jfoyp4fe4c9-4{list-style-type:none}.lst-kix_ova5lqw9xm1i-5>li:before{content:"\0025a0   "}ul.lst-kix_jfoyp4fe4c9-3{list-style-type:none}ul.lst-kix_jfoyp4fe4c9-2{list-style-type:none}ul.lst-kix_jfoyp4fe4c9-1{list-style-type:none}ul.lst-kix_jfoyp4fe4c9-0{list-style-type:none}.lst-kix_xwb2xgsmkc7s-4>li:before{content:"\0025cb   "}.lst-kix_z7kpv721mdo1-0>li:before{content:"\0025cf   "}.lst-kix_z7kpv721mdo1-2>li:before{content:"\0025a0   "}.lst-kix_xwb2xgsmkc7s-6>li:before{content:"\0025cf   "}ul.lst-kix_z7kpv721mdo1-7{list-style-type:none}ul.lst-kix_z7kpv721mdo1-6{list-style-type:none}.lst-kix_i97szj75x3di-8>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-8,lower-roman) ". "}ul.lst-kix_z7kpv721mdo1-8{list-style-type:none}.lst-kix_xwb2xgsmkc7s-8>li:before{content:"\0025a0   "}.lst-kix_lryd0r56zvdy-6>li:before{content:"\0025cf   "}.lst-kix_lryd0r56zvdy-8>li:before{content:"\0025a0   "}ul.lst-kix_jfoyp4fe4c9-8{list-style-type:none}.lst-kix_d8exnt1kyoig-7>li:before{content:"\0025cb   "}.lst-kix_khhpt4pbx05-5>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-5,lower-roman) ". "}.lst-kix_lryd0r56zvdy-2>li:before{content:"\0025a0   "}.lst-kix_lryd0r56zvdy-4>li:before{content:"\0025cb   "}.lst-kix_d8exnt1kyoig-3>li:before{content:"\0025cf   "}ul.lst-kix_z7kpv721mdo1-3{list-style-type:none}ul.lst-kix_z7kpv721mdo1-2{list-style-type:none}ul.lst-kix_z7kpv721mdo1-5{list-style-type:none}ul.lst-kix_z7kpv721mdo1-4{list-style-type:none}.lst-kix_d8exnt1kyoig-1>li:before{content:"\0025cb   "}.lst-kix_khhpt4pbx05-7>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-7,lower-latin) ". "}ul.lst-kix_z7kpv721mdo1-1{list-style-type:none}.lst-kix_i97szj75x3di-3>li{counter-increment:lst-ctn-kix_i97szj75x3di-3}ul.lst-kix_z7kpv721mdo1-0{list-style-type:none}.lst-kix_firddmuqun3h-7>li:before{content:"\0025cb   "}.lst-kix_khhpt4pbx05-1>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-1,lower-latin) ". "}.lst-kix_xwb2xgsmkc7s-2>li:before{content:"\0025a0   "}.lst-kix_firddmuqun3h-5>li:before{content:"\0025a0   "}.lst-kix_lryd0r56zvdy-0>li:before{content:"\0025cf   "}ol.lst-kix_khhpt4pbx05-3.start{counter-reset:lst-ctn-kix_khhpt4pbx05-3 0}.lst-kix_d8exnt1kyoig-5>li:before{content:"\0025a0   "}.lst-kix_khhpt4pbx05-3>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-3,decimal) ". "}.lst-kix_xwb2xgsmkc7s-0>li:before{content:"\0025cf   "}.lst-kix_firddmuqun3h-3>li:before{content:"\0025cf   "}.lst-kix_rysblip21ldc-7>li:before{content:"\0025cb   "}ol.lst-kix_khhpt4pbx05-2.start{counter-reset:lst-ctn-kix_khhpt4pbx05-2 0}.lst-kix_firddmuqun3h-1>li:before{content:"\0025cb   "}.lst-kix_khhpt4pbx05-6>li{counter-increment:lst-ctn-kix_khhpt4pbx05-6}ol.lst-kix_i97szj75x3di-5.start{counter-reset:lst-ctn-kix_i97szj75x3di-5 0}.lst-kix_khhpt4pbx05-0>li{counter-increment:lst-ctn-kix_khhpt4pbx05-0}ul.lst-kix_ova5lqw9xm1i-7{list-style-type:none}ul.lst-kix_ova5lqw9xm1i-6{list-style-type:none}ul.lst-kix_ova5lqw9xm1i-8{list-style-type:none}ul.lst-kix_ova5lqw9xm1i-3{list-style-type:none}ul.lst-kix_ova5lqw9xm1i-2{list-style-type:none}ul.lst-kix_ova5lqw9xm1i-5{list-style-type:none}ol.lst-kix_i97szj75x3di-0.start{counter-reset:lst-ctn-kix_i97szj75x3di-0 0}ul.lst-kix_ova5lqw9xm1i-4{list-style-type:none}.lst-kix_arwys544g8ik-7>li:before{content:"\0025cb   "}.lst-kix_arwys544g8ik-6>li:before{content:"\0025cf   "}.lst-kix_khhpt4pbx05-4>li{counter-increment:lst-ctn-kix_khhpt4pbx05-4}.lst-kix_arwys544g8ik-4>li:before{content:"\0025cb   "}.lst-kix_arwys544g8ik-3>li:before{content:"\0025cf   "}.lst-kix_arwys544g8ik-5>li:before{content:"\0025a0   "}.lst-kix_arwys544g8ik-0>li:before{content:"\0025cf   "}.lst-kix_arwys544g8ik-1>li:before{content:"\0025cb   "}ul.lst-kix_ova5lqw9xm1i-1{list-style-type:none}ul.lst-kix_ova5lqw9xm1i-0{list-style-type:none}.lst-kix_arwys544g8ik-2>li:before{content:"\0025a0   "}.lst-kix_i97szj75x3di-7>li{counter-increment:lst-ctn-kix_i97szj75x3di-7}ul.lst-kix_arwys544g8ik-0{list-style-type:none}.lst-kix_i97szj75x3di-5>li{counter-increment:lst-ctn-kix_i97szj75x3di-5}.lst-kix_khhpt4pbx05-2>li{counter-increment:lst-ctn-kix_khhpt4pbx05-2}ol.lst-kix_i97szj75x3di-6.start{counter-reset:lst-ctn-kix_i97szj75x3di-6 0}ul.lst-kix_arwys544g8ik-3{list-style-type:none}ul.lst-kix_arwys544g8ik-4{list-style-type:none}ul.lst-kix_arwys544g8ik-1{list-style-type:none}ul.lst-kix_arwys544g8ik-2{list-style-type:none}ul.lst-kix_arwys544g8ik-7{list-style-type:none}ul.lst-kix_arwys544g8ik-8{list-style-type:none}ul.lst-kix_arwys544g8ik-5{list-style-type:none}ul.lst-kix_arwys544g8ik-6{list-style-type:none}.lst-kix_firddmuqun3h-8>li:before{content:"\0025a0   "}ol.lst-kix_khhpt4pbx05-4.start{counter-reset:lst-ctn-kix_khhpt4pbx05-4 0}ul.lst-kix_rysblip21ldc-4{list-style-type:none}ul.lst-kix_rysblip21ldc-5{list-style-type:none}ul.lst-kix_d8exnt1kyoig-8{list-style-type:none}ul.lst-kix_rysblip21ldc-2{list-style-type:none}ul.lst-kix_d8exnt1kyoig-7{list-style-type:none}ul.lst-kix_rysblip21ldc-3{list-style-type:none}ul.lst-kix_d8exnt1kyoig-6{list-style-type:none}ul.lst-kix_rysblip21ldc-8{list-style-type:none}ul.lst-kix_rysblip21ldc-6{list-style-type:none}ul.lst-kix_rysblip21ldc-7{list-style-type:none}ul.lst-kix_firddmuqun3h-1{list-style-type:none}ul.lst-kix_firddmuqun3h-2{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-0{list-style-type:none}ul.lst-kix_firddmuqun3h-3{list-style-type:none}ul.lst-kix_firddmuqun3h-4{list-style-type:none}ul.lst-kix_firddmuqun3h-5{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-3{list-style-type:none}ul.lst-kix_firddmuqun3h-6{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-4{list-style-type:none}ul.lst-kix_firddmuqun3h-7{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-1{list-style-type:none}ul.lst-kix_firddmuqun3h-8{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-2{list-style-type:none}ul.lst-kix_rysblip21ldc-0{list-style-type:none}.lst-kix_khhpt4pbx05-8>li{counter-increment:lst-ctn-kix_khhpt4pbx05-8}ul.lst-kix_rysblip21ldc-1{list-style-type:none}ul.lst-kix_firddmuqun3h-0{list-style-type:none}ul.lst-kix_lryd0r56zvdy-2{list-style-type:none}ul.lst-kix_lryd0r56zvdy-3{list-style-type:none}ul.lst-kix_lryd0r56zvdy-0{list-style-type:none}ul.lst-kix_lryd0r56zvdy-1{list-style-type:none}ul.lst-kix_lryd0r56zvdy-6{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-7{list-style-type:none}ul.lst-kix_lryd0r56zvdy-7{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-8{list-style-type:none}ul.lst-kix_lryd0r56zvdy-4{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-5{list-style-type:none}ul.lst-kix_lryd0r56zvdy-5{list-style-type:none}ul.lst-kix_kfnp1fjph7ti-6{list-style-type:none}ol.lst-kix_i97szj75x3di-1.start{counter-reset:lst-ctn-kix_i97szj75x3di-1 0}ul.lst-kix_lryd0r56zvdy-8{list-style-type:none}ol.lst-kix_i97szj75x3di-6{list-style-type:none}ol.lst-kix_i97szj75x3di-5{list-style-type:none}.lst-kix_csthsd1cfrto-1>li:before{content:"\0025cb   "}ol.lst-kix_i97szj75x3di-8{list-style-type:none}ol.lst-kix_i97szj75x3di-7{list-style-type:none}ol.lst-kix_i97szj75x3di-2{list-style-type:none}ol.lst-kix_i97szj75x3di-1{list-style-type:none}.lst-kix_csthsd1cfrto-0>li:before{content:"\0025cf   "}ol.lst-kix_i97szj75x3di-4{list-style-type:none}ol.lst-kix_i97szj75x3di-3{list-style-type:none}ol.lst-kix_i97szj75x3di-0{list-style-type:none}ul.lst-kix_csthsd1cfrto-2{list-style-type:none}.lst-kix_kfnp1fjph7ti-5>li:before{content:"\0025a0   "}ol.lst-kix_khhpt4pbx05-8.start{counter-reset:lst-ctn-kix_khhpt4pbx05-8 0}ul.lst-kix_csthsd1cfrto-1{list-style-type:none}ul.lst-kix_csthsd1cfrto-0{list-style-type:none}.lst-kix_jfoyp4fe4c9-0>li:before{content:"\0025cf   "}.lst-kix_jfoyp4fe4c9-2>li:before{content:"\0025a0   "}.lst-kix_kfnp1fjph7ti-3>li:before{content:"\0025cf   "}.lst-kix_kfnp1fjph7ti-7>li:before{content:"\0025cb   "}.lst-kix_kfnp1fjph7ti-1>li:before{content:"\0025cb   "}.lst-kix_i97szj75x3di-0>li{counter-increment:lst-ctn-kix_i97szj75x3di-0}.lst-kix_jfoyp4fe4c9-6>li:before{content:"\0025cf   "}.lst-kix_csthsd1cfrto-2>li:before{content:"\0025a0   "}.lst-kix_csthsd1cfrto-4>li:before{content:"\0025cb   "}.lst-kix_i97szj75x3di-5>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-5,lower-roman) ". "}.lst-kix_rysblip21ldc-2>li:before{content:"\0025a0   "}.lst-kix_i97szj75x3di-1>li{counter-increment:lst-ctn-kix_i97szj75x3di-1}.lst-kix_ova5lqw9xm1i-2>li:before{content:"\0025a0   "}.lst-kix_rysblip21ldc-4>li:before{content:"\0025cb   "}.lst-kix_csthsd1cfrto-8>li:before{content:"\0025a0   "}ul.lst-kix_csthsd1cfrto-8{list-style-type:none}.lst-kix_i97szj75x3di-1>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-1,lower-latin) ". "}.lst-kix_i97szj75x3di-3>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-3,decimal) ". "}ul.lst-kix_csthsd1cfrto-7{list-style-type:none}.lst-kix_jfoyp4fe4c9-4>li:before{content:"\0025cb   "}ul.lst-kix_csthsd1cfrto-6{list-style-type:none}ul.lst-kix_csthsd1cfrto-5{list-style-type:none}.lst-kix_csthsd1cfrto-6>li:before{content:"\0025cf   "}ul.lst-kix_csthsd1cfrto-4{list-style-type:none}.lst-kix_ova5lqw9xm1i-4>li:before{content:"\0025cb   "}ul.lst-kix_csthsd1cfrto-3{list-style-type:none}.lst-kix_ova5lqw9xm1i-6>li:before{content:"\0025cf   "}.lst-kix_ova5lqw9xm1i-8>li:before{content:"\0025a0   "}.lst-kix_rysblip21ldc-0>li:before{content:"\0025cf   "}.lst-kix_z7kpv721mdo1-1>li:before{content:"\0025cb   "}.lst-kix_xwb2xgsmkc7s-5>li:before{content:"\0025a0   "}.lst-kix_i97szj75x3di-7>li:before{content:"" counter(lst-ctn-kix_i97szj75x3di-7,lower-latin) ". "}.lst-kix_jfoyp4fe4c9-8>li:before{content:"\0025a0   "}.lst-kix_d8exnt1kyoig-0>li:before{content:"\0025cf   "}.lst-kix_lryd0r56zvdy-7>li:before{content:"\0025cb   "}ul.lst-kix_d8exnt1kyoig-1{list-style-type:none}ul.lst-kix_d8exnt1kyoig-0{list-style-type:none}.lst-kix_xwb2xgsmkc7s-7>li:before{content:"\0025cb   "}ul.lst-kix_d8exnt1kyoig-5{list-style-type:none}ul.lst-kix_d8exnt1kyoig-4{list-style-type:none}ul.lst-kix_d8exnt1kyoig-3{list-style-type:none}ul.lst-kix_d8exnt1kyoig-2{list-style-type:none}.lst-kix_lryd0r56zvdy-3>li:before{content:"\0025cf   "}.lst-kix_khhpt4pbx05-6>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-6,decimal) ". "}.lst-kix_lryd0r56zvdy-1>li:before{content:"\0025cb   "}.lst-kix_lryd0r56zvdy-5>li:before{content:"\0025a0   "}.lst-kix_khhpt4pbx05-4>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-4,lower-latin) ". "}.lst-kix_khhpt4pbx05-8>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-8,lower-roman) ". "}.lst-kix_d8exnt1kyoig-4>li:before{content:"\0025cb   "}.lst-kix_d8exnt1kyoig-8>li:before{content:"\0025a0   "}.lst-kix_d8exnt1kyoig-2>li:before{content:"\0025a0   "}.lst-kix_xwb2xgsmkc7s-3>li:before{content:"\0025cf   "}.lst-kix_firddmuqun3h-6>li:before{content:"\0025cf   "}.lst-kix_khhpt4pbx05-0>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-0,decimal) ". "}.lst-kix_khhpt4pbx05-2>li:before{content:"" counter(lst-ctn-kix_khhpt4pbx05-2,lower-roman) ". "}.lst-kix_xwb2xgsmkc7s-1>li:before{content:"\0025cb   "}.lst-kix_firddmuqun3h-2>li:before{content:"\0025a0   "}.lst-kix_firddmuqun3h-4>li:before{content:"\0025cb   "}.lst-kix_d8exnt1kyoig-6>li:before{content:"\0025cf   "}ul.lst-kix_xwb2xgsmkc7s-8{list-style-type:none}ul.lst-kix_e63skeaub1i7-4{list-style-type:none}ul.lst-kix_xwb2xgsmkc7s-7{list-style-type:none}ul.lst-kix_e63skeaub1i7-3{list-style-type:none}ul.lst-kix_xwb2xgsmkc7s-6{list-style-type:none}ul.lst-kix_e63skeaub1i7-2{list-style-type:none}ul.lst-kix_xwb2xgsmkc7s-5{list-style-type:none}ul.lst-kix_e63skeaub1i7-1{list-style-type:none}.lst-kix_firddmuqun3h-0>li:before{content:"\0025cf   "}ul.lst-kix_e63skeaub1i7-0{list-style-type:none}.lst-kix_rysblip21ldc-6>li:before{content:"\0025cf   "}li.li-bullet-0:before{margin-left:-18pt;white-space:nowrap;display:inline-block;min-width:18pt}.lst-kix_ova5lqw9xm1i-0>li:before{content:"\0025cf   "}.lst-kix_rysblip21ldc-8>li:before{content:"\0025a0   "}.lst-kix_khhpt4pbx05-3>li{counter-increment:lst-ctn-kix_khhpt4pbx05-3}ul.lst-kix_xwb2xgsmkc7s-0{list-style-type:none}.lst-kix_arwys544g8ik-8>li:before{content:"\0025a0   "}ul.lst-kix_xwb2xgsmkc7s-4{list-style-type:none}ul.lst-kix_xwb2xgsmkc7s-3{list-style-type:none}ul.lst-kix_xwb2xgsmkc7s-2{list-style-type:none}ul.lst-kix_xwb2xgsmkc7s-1{list-style-type:none}ol{margin:0;padding:0}table td,table th{padding:0}.c39{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:95.2pt;border-top-color:#333333;border-bottom-style:solid}.c29{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:170.2pt;border-top-color:#333333;border-bottom-style:solid}.c37{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:154.8pt;border-top-color:#333333;border-bottom-style:solid}.c45{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:110pt;border-top-color:#333333;border-bottom-style:solid}.c14{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:188.8pt;border-top-color:#333333;border-bottom-style:solid}.c20{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:343pt;border-top-color:#333333;border-bottom-style:solid}.c35{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:108.8pt;border-top-color:#333333;border-bottom-style:solid}.c9{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:135pt;border-top-color:#333333;border-bottom-style:solid}.c13{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:66.8pt;border-top-color:#333333;border-bottom-style:solid}.c30{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:111.5pt;border-top-color:#333333;border-bottom-style:solid}.c18{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:132.8pt;border-top-color:#333333;border-bottom-style:solid}.c36{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:124.5pt;border-top-color:#333333;border-bottom-style:solid}.c19{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:105.8pt;border-top-color:#333333;border-bottom-style:solid}.c12{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:81.8pt;border-top-color:#333333;border-bottom-style:solid}.c43{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:84.5pt;border-top-color:#333333;border-bottom-style:solid}.c16{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:159.8pt;border-top-color:#333333;border-bottom-style:solid}.c21{border-right-style:solid;padding:5pt 5pt 5pt 5pt;border-bottom-color:#333333;border-top-width:1pt;border-right-width:1pt;border-left-color:#333333;vertical-align:top;border-right-color:#333333;border-left-width:1pt;border-top-style:solid;border-left-style:solid;border-bottom-width:1pt;width:121.8pt;border-top-color:#333333;border-bottom-style:solid}.c10{margin-left:36pt;padding-top:12pt;padding-left:0pt;padding-bottom:12pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c4{color:#000000;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:13pt;font-family:"Arial";font-style:normal}.c1{color:#188038;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Roboto Mono";font-style:normal}.c7{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c0{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left;height:11pt}.c52{padding-top:18pt;padding-bottom:4pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c2{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c27{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:center}.c41{padding-top:14pt;padding-bottom:4pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c15{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:right}.c8{font-weight:400;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c53{color:#000000;text-decoration:none;vertical-align:super;font-size:11pt;font-family:"Arial"}.c56{color:#000000;text-decoration:none;vertical-align:sub;font-size:11pt;font-family:"Arial"}.c24{color:#000000;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial"}.c55{color:#000000;text-decoration:none;vertical-align:baseline;font-size:15pt;font-family:"Arial"}.c5{text-decoration-skip-ink:none;-webkit-text-decoration-skip:none;color:#1155cc;text-decoration:underline}.c25{border-spacing:0;border-collapse:collapse;margin-right:auto}.c54{margin-left:auto;border-spacing:0;border-collapse:collapse;margin-right:auto}.c3{font-family:"Roboto Mono";color:#188038;font-weight:400}.c28{background-color:#ffffff;max-width:468pt;padding:72pt 72pt 72pt 72pt}.c32{color:inherit;text-decoration:inherit}.c11{padding:0;margin:0}.c46{margin-left:144pt;text-indent:36pt}.c47{margin-left:36pt;text-indent:36pt}.c40{margin-left:72pt;text-indent:36pt}.c44{margin-left:108pt;text-indent:36pt}.c34{font-style:normal}.c38{height:12.2pt}.c33{font-weight:400}.c48{height:36.8pt}.c51{text-indent:36pt}.c26{height:25.8pt}.c49{height:45.4pt}.c50{margin-left:72pt}.c42{height:52.8pt}.c23{color:#188038}.c6{font-style:italic}.c31{height:39.2pt}.c22{font-weight:700}.c17{height:66.2pt}.title{padding-top:0pt;color:#000000;font-size:26pt;padding-bottom:3pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:20pt;color:#000000;font-size:20pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:18pt;color:#000000;font-size:16pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}</style></head><body class="c28 doc-content"><h2 class="c52" id="h.fa398hho6yc2"><span class="c34 c22 c55">Thermo QuantStudio Real-Time PCR Software Adapter Requirements</span></h2><p class="c2"><span class="c7">Revision 05</span></p><p class="c2"><span class="c7">Date: 2024-01-19</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">The Thermo QuantStudio adapter is intended to convert text file (.txt) exports from QuantStudio RealTime PCR Software to the qPCR Allotrope Simple Model (ASM). Thermo QuantStudio PCR systems are quantitative PCR (qPCR) - also known as real-time PCR - and are used to detect the presence and absence of specific DNA oligonucleotides. The adapter is expected to parse data from the instrument pertaining to the amplification of target amplicons and associated results fields.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">The adapter is expected to:</span></p><ul class="c11 lst-kix_jfoyp4fe4c9-0 start"><li class="c10 li-bullet-0"><span class="c7">Watch a designated directory for .txt files matching the pattern of QuantStudio exports</span></li><li class="c10 li-bullet-0"><span class="c7">Convert the exported .txt file to an Allotrope .json file</span></li><li class="c10 li-bullet-0"><span class="c7">Return the original .txt file and the converted ASM .json file to a designated S3 bucket</span></li></ul><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">The adapter is expected to produce the following output based on the provided input:</span></p><p class="c0"><span class="c7"></span></p><ul class="c11 lst-kix_z7kpv721mdo1-0 start"><li class="c10 li-bullet-0"><span>Input:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1q_2Li_SKAIfh_CSupUCY0KrdIwkE3pHT/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185211923&amp;usg=AOvVaw3nia9hqYaGno_G8wGOnw_c">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1q_2Li_SKAIfh_CSupUCY0KrdIwkE3pHT/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185212272&amp;usg=AOvVaw3XElcigUnaYLY3Ak-4A_b7">Thermo_QuantStudio_example01_instrumentOutput.txt</a></span></li><li class="c10 li-bullet-0"><span>Output:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1rBjm3oyNdLZhGNh7fIAG5aodSnW7GV-M/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185212560&amp;usg=AOvVaw16aPrV242JdOZnsaKoy2Eo">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1rBjm3oyNdLZhGNh7fIAG5aodSnW7GV-M/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185212727&amp;usg=AOvVaw1C9wkzv-v1obTLL0YDrNOp">Thermo_QuantStudio_example01_asm_rev02.json</a></span></li></ul><p class="c2"><span class="c24 c33 c6">For this example, the output only contains the data for (6) wells [A1, A2, A3, B1, B2, B3] present in the input example. These wells represent data for (3) samples run with (2) target DNA, one a control and the other a gene of interest. The example data cubes only contain the first (3).</span></p><ul class="c11 lst-kix_d8exnt1kyoig-0 start"><li class="c10 li-bullet-0"><span>Input:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1OxeJ9n7MbPE5tbEcfuj2O65rGoOCbGqx/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185213117&amp;usg=AOvVaw3_xyNcnrhAeENUjPUdZI2n">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1OxeJ9n7MbPE5tbEcfuj2O65rGoOCbGqx/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185213293&amp;usg=AOvVaw3ixNq9A8Ujk9wgtz222Qah">Thermo_QuantStudio_example02_instrumentOutput.txt</a></span></li><li class="c10 li-bullet-0"><span>Output:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1I3SUkzGlRZSea0Dk1ZogZOgg2ChpH2U5/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185213662&amp;usg=AOvVaw0W7IIaQkRPBNduM48PX6Vv">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1I3SUkzGlRZSea0Dk1ZogZOgg2ChpH2U5/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185213822&amp;usg=AOvVaw0BOJAv6IrWkWnLMPG05grq">Thermo_QuantStudio_example02_asm.json</a></span></li></ul><p class="c2"><span class="c24 c6 c33">For this example, the output only contains the (2) samples present in the input examples, these are also from non-consecutive wells A1 and A12. The datacubes only contain the first (3) entries</span></p><p class="c0"><span class="c7"></span></p><ul class="c11 lst-kix_csthsd1cfrto-0 start"><li class="c10 li-bullet-0"><span>Input:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1YvmQm7U-GoY0VW6voJb_xiKQfJG4aGWi/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185214427&amp;usg=AOvVaw2oZnFSCTvhBbwSkcuPty8a">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1YvmQm7U-GoY0VW6voJb_xiKQfJG4aGWi/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185214605&amp;usg=AOvVaw1hkqnKQLXykfTwv59ICwWr">Thermo_QuantStudio_example03_instrumentOutput.txt</a></span></li><li class="c10 li-bullet-0"><span>Output:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1id11q9at751_vX1ogPTQI3lO3qA_yzIt/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185214892&amp;usg=AOvVaw1wzkfEN8l5KdQbUUZ3zMDX">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1id11q9at751_vX1ogPTQI3lO3qA_yzIt/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185215080&amp;usg=AOvVaw2LwCvHsgMwSAKE7dbzNfXf">Thermo_QuantStudio_example03_asm.json</a></span></li></ul><p class="c2"><span class="c24 c33 c6">This example depicts a Genotyping Experiment Type showing how the data from each Allele should be segmented to separate measurement documents (discussed below). For brevity, this example only contains place holders for the datacubes, but the actual output is expected to contain populated datacubes.</span></p><p class="c0"><span class="c7"></span></p><ul class="c11 lst-kix_kfnp1fjph7ti-0 start"><li class="c10 li-bullet-0"><span>Input:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1ysQ-mIF3VtV-8ol2F4W6CB9YencbSKIf/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185215609&amp;usg=AOvVaw1CE3dVZQshfXzwcdgfX-ib">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1ysQ-mIF3VtV-8ol2F4W6CB9YencbSKIf/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185215813&amp;usg=AOvVaw2Bnz1fBkj1tXuTKKTMs74R">Thermo_QuanStudio_example04_instrumentOutput.txt</a></span></li></ul><p class="c2"><span class="c24 c33 c6">This example depicts a case in which the Calculated Data Document for Ct Mean would reference several Measurement Documents (discussed below).</span></p><p class="c0"><span class="c7"></span></p><ul class="c11 lst-kix_arwys544g8ik-0 start"><li class="c10 li-bullet-0"><span>Input:</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1tvKvMYFAaIzloJJwe0Xk2tI7D1ZctT7t/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185216291&amp;usg=AOvVaw3-iGmvrO5L4XzYgYoxmRuz">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1tvKvMYFAaIzloJJwe0Xk2tI7D1ZctT7t/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185216454&amp;usg=AOvVaw3mDh2JZ3c2PRS0xUO1IsqT">Thermo_QuantStudio_example05_instrumentOutput.txt</a></span></li></ul><p class="c2"><span>This example depicts a case of a Comparative Ct experiment type for which Ct Mean would reference several Measurement Documents, in addition to requiring Calculated Data Documents for </span><span class="c6">Delta Ct Mean, Delta Delta Ct, </span><span>and </span><span class="c24 c33 c6">RQ</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">The ASM for qPCR is detailed in the following:</span></p><ul class="c11 lst-kix_firddmuqun3h-0 start"><li class="c10 li-bullet-0"><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://docs.google.com/spreadsheets/d/120I7OyLvrkHdM0hQA3sfcAg40ZaGZu41/edit?usp%3Ddrive_link%26ouid%3D111356359899923803205%26rtpof%3Dtrue%26sd%3Dtrue&amp;sa=D&amp;source=editors&amp;ust=1715829185217033&amp;usg=AOvVaw1ScE-hB1Ba6GisbsV6xHg8">diagram of qPCR ASM rev04</a></span></li><li class="c10 li-bullet-0"><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1xVlLbxejUSnITq_BjI9t25SFDblXXyTr/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185217305&amp;usg=AOvVaw3KFR7Ir2qXh1WR8tGn1PbF">simplified qPCR JSON schema rev03</a></span></li></ul><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.gb9t8u49krbn"><span class="c4">Required User Actions</span></h3><p class="c2"><span>The required user actions are based instructions provided in the</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1nmjX-lO9C4QQ-_qHOBvD0a7HkkAmQLyh/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185217843&amp;usg=AOvVaw0cO1K3WWeyXP8hyFjPXeex">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1nmjX-lO9C4QQ-_qHOBvD0a7HkkAmQLyh/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185218057&amp;usg=AOvVaw2q2Kqk8Ocp-aoNe8604WCS">QuantStudio Real-Time PCR Software Users Manual</a></span><span class="c5">&nbsp;</span><span class="c7">(PN 4489822, Rev A).</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">The user is expected to enter Sample Name values within the QuantStudio UI that conform to entity identifiers within Benchling, either existing entity IDs or entities they wish to register upon processing of results. The user may do this by manually entering Sample Names or Importing a Sample Definition File (see p.59 IFU).</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The user is expect to generate a .txt export from the </span><span class="c6">Export</span><span class="c7">&nbsp;view within the QuantStudio UI.</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 230.67px;"><img alt="QuantStudio_exportScreen_exportsidebar.png" src="images/image6.png" style="width: 624.00px; height: 230.67px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c7">The user can select the data to include within the export by selecting the individual tabs.</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 230.67px;"><img alt="QuantStudio_exportScreen_tabs.png" src="images/image1.png" style="width: 624.00px; height: 230.67px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span>The adapter is able to process the data within the: </span><span class="c6">Sample Setup</span><span>, </span><span class="c6">Amplification</span><span>, </span><span class="c6">Multicomponent</span><span>, and </span><span class="c6">Results</span><span class="c7">&nbsp;tabs</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 53.33px;"><img alt="QuantStudio_exportScreen_tabs_crop.png" src="images/image5.png" style="width: 624.00px; height: 53.33px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span>The user is expected keep the default option to </span><span class="c24 c33 c6">Export Data To One File</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 230.67px;"><img alt="QuantStudio_exportScreen_singleFileExport.png" src="images/image7.png" style="width: 624.00px; height: 230.67px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c7">The user should select .txt as the file type. Processing of the .xls and .xlsx formats is not currently supported by the adapter.</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 230.67px;"><img alt="QuantStudio_exportScreen_filetype.png" src="images/image3.png" style="width: 624.00px; height: 230.67px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span>The user should set the </span><span class="c6">Export File Location</span><span>&nbsp;to be the target directory specified during </span><span class="c6">Connection</span><span class="c7">&nbsp;configuration.</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 230.67px;"><img alt="QuantStudio_exportScreen_filelocation.png" src="images/image4.png" style="width: 624.00px; height: 230.67px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span>In order to insure that the exported files are consistently placed in the target directory for the </span><span class="c6">Connection</span><span class="c7">, the default export location can be set within the QuantStudio preferences menu.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 506.67px;"><img alt="QuantStudio_preferences_exportFolder.png" src="images/image2.png" style="width: 624.00px; height: 506.67px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><h3 class="c41" id="h.hhh63fwuwg7m"><span class="c4">Generation of the ASM file</span></h3><p class="c2"><span class="c7">The Allotrope ASM for qPCR has the following structure</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c1">qPCR Aggregate Document</span></p><p class="c2"><span class="c23">&#11153; </span><span class="c1">Device System Document</span></p><p class="c2"><span class="c23">&#11153; </span><span class="c1">Data System Document</span></p><p class="c2"><span class="c23">&#11153; </span><span class="c1">qPCR Document</span></p><p class="c2 c51"><span class="c1">&#11153; Measurement Aggregate Document</span></p><p class="c2 c47"><span class="c23">&#11153; </span><span class="c1">Measurement Document(s)</span></p><p class="c2 c40"><span class="c23">&#11153; </span><span class="c1">Sample Document</span></p><p class="c2 c40"><span class="c23">&#11153; </span><span class="c1">Device Control Aggregate Document</span></p><p class="c2 c44"><span class="c23">&#11153; </span><span class="c1">Device Control Document(s)</span></p><p class="c2 c40"><span class="c23">&#11153; </span><span class="c1">Processed Data Aggregate Document</span></p><p class="c2 c44"><span class="c23">&#11153; </span><span class="c1">Processed Data Document</span></p><p class="c2 c46"><span class="c23">&#11153; </span><span class="c1">Data Processing Document</span></p><p class="c2"><span class="c23">&#11153; </span><span class="c1">Calculated Data Aggregate Document</span></p><p class="c2 c51"><span class="c23">&#11153; </span><span class="c1">Calculated Data Document</span></p><p class="c2 c50"><span class="c23">&#11153; </span><span class="c1">Data Processing Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">The QuantStudio export file has a number of &lsquo;blocks&rsquo; which contain relevant measurement and metadata required for the ASM. The &lsquo;blocks&rsquo; follow the pattern of:</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c6">header </span><span class="c7">- detailing device, file, and experiment information</span></p><p class="c2"><span class="c24 c33 c6">sample setup</span></p><p class="c2"><span class="c24 c33 c6">raw data</span></p><p class="c2"><span class="c24 c33 c6">amplification data</span></p><p class="c2"><span class="c24 c33 c6">multicomponent data</span></p><p class="c2"><span class="c24 c33 c6">results</span></p><p class="c2"><span class="c24 c33 c6">melt curve raw data</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.7wfwmnxx4tu6"><span class="c4">qPCR Aggregate Document</span></h3><p class="c2"><span>The </span><span class="c3">qPCR Aggregate Document</span><span>&nbsp;will contain the </span><span class="c3">Device System Document</span><span>&nbsp;, </span><span class="c3">Data System Document</span><span>, </span><span class="c3">qPCR Document</span><span>, and where appropriate a</span><span class="c3">&nbsp;Calculated Data Aggregate Document</span><span>&nbsp;for recording processed data derived from multiple </span><span class="c3">qPCR Documents</span><span class="c7">.</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.meqh9nxv4aa"><span class="c4">Device System Document</span></h3><p class="c2"><span>The </span><span class="c3">device system document</span><span class="c7">&nbsp;within the ASM is intended to capture metadata about the instrument from which the data was obtained.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The details for </span><span class="c3">device identifier</span><span>, </span><span class="c3">model number</span><span>, and </span><span class="c3">device serial number</span><span>&nbsp;are all provided within the </span><span class="c6">header </span><span class="c7">block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">device identifier</span><span>&nbsp;is recorded within the </span><span class="c6">Instrument Name</span><span>&nbsp;field. While this field defaults to the instrument serial number, it should be recorded within the </span><span class="c3">device identifier</span><span class="c7">&nbsp;field even when the two are repeated.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>Meanwhile, the </span><span class="c3">model number</span><span>&nbsp;is provided within the </span><span class="c6">Instrument Type </span><span>field and </span><span class="c3">device serial number</span><span>&nbsp;is provided within the </span><span class="c6">Instrument Serial Number </span><span class="c7">field.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>All other values within the </span><span class="c3">Device System Document</span><span class="c7">&nbsp;can be ignored at this time.</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.yz20pdz5oltu"><span class="c4">Data System Document</span></h3><p class="c2"><span>The </span><span class="c3">data system document</span><span class="c7">&nbsp;within the ASM is intended to capture information about the instrument file, its location, source of generation, the adapter/converter used to generate the ASM file.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">data system instance identifier</span><span class="c7">&nbsp;is intended to reference the computer/workstation/file-store on which the instrument file resides. This information is not included in the export, but if possible should be obtained by the adapter querying the identifier of the local system on which it is installed.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">file name</span><span class="c7">&nbsp;is the name of the instrument file which the adapter is using as input.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">UNC path</span><span class="c7">&nbsp;should be the complete file path of the instrument file.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">software name</span><span>&nbsp;and </span><span class="c3">software version</span><span>&nbsp;refer to the application that generated the instrument file. This information is not included in the QuantStudio export, therefore &lsquo;QuantStudio Real-Time PCR Software&rsquo; can be used as the default name for </span><span class="c3">software name</span><span>&nbsp;while </span><span class="c3">software version</span><span class="c7">&nbsp;can be excluded from the ASM at this time.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">ASM converter name</span><span>&nbsp;and </span><span class="c3">ASM converter version</span><span>&nbsp;refers to the </span><span class="c6">the adapter being described in this requirements document</span><span class="c7">&nbsp;used generate the ASM file from the instrument file. Therefore, the adapter should populate these fields with its own name and version.</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.lh1knyngdp1a"><span class="c4">qPCR Document</span></h3><p class="c2"><span>The </span><span class="c3">qPCR Document</span><span>&nbsp;is an example of a </span><span class="c6">technique document</span><span>&nbsp;in the Allotrope parlance, and meant to represent each time the qPCR technique is perform. In this framework the qPCR technique is performed for each plate well in the QuantStudio export - as qPCR instruments typically rely on multi well plates. The blocks within the QuantStudio export will reference </span><span class="c6">Well </span><span>or </span><span class="c6">Well Position</span><span>&nbsp;(depending on the block) - these should then correspond to individual </span><span class="c3">qPCR Documents</span><span class="c7">.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">qPCR Document</span><span>&nbsp;contains a </span><span class="c3">Measurement Aggregate Document</span><span>, details about </span><span class="c3">analyst</span><span>&nbsp;- the individual who executed the procedure, and the </span><span class="c3">submitter</span><span class="c7">&nbsp;- the individual who requested the procedure and submitted samples to the analyst.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The entry for </span><span class="c3">analyst</span><span>&nbsp;is recorded under the </span><span class="c6">Experiment User Name</span><span class="c7">&nbsp;in the header block. If no entry is provided - &lsquo;NA&rsquo; - then this field can be ignored.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>There is no entry corresponding to </span><span class="c3">submitter</span><span class="c7">&nbsp;within the QuantStudio export.</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.9mybebxs2re6"><span class="c4">Measurement Aggregate Document</span></h3><p class="c2"><span>The </span><span class="c3">Measurement Aggregate Document</span><span>&nbsp;contains details about the </span><span class="c3">analytical method identifier</span><span>, </span><span class="c3">experimental data identifier</span><span>, </span><span class="c3">experiment type</span><span>, </span><span class="c3">container type</span><span>, </span><span class="c3">plate well count</span><span>, </span><span class="c3">well volume</span><span>, and details about the individual measurements within a </span><span class="c3">Measurement Document</span><span class="c7">&nbsp;array.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The QuantStudio export does not provide a field corresponding to </span><span class="c3">analytical method identifier</span><span class="c7">, so this entry can be ignored.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">experimental data identifier</span><span>&nbsp;should be taken from the </span><span class="c6">Experiment Name</span><span>&nbsp;within the </span><span class="c6">header </span><span class="c7">block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">experiment type</span><span>&nbsp;is recorded within the </span><span class="c6">Experiment Type</span><span>&nbsp;field within the </span><span class="c6">header </span><span>block. The record of </span><span class="c6">Experiment Type</span><span class="c7">&nbsp;should correspond to one of the following:</span></p><ul class="c11 lst-kix_lryd0r56zvdy-0 start"><li class="c10 li-bullet-0"><span class="c7">Standard Curve</span></li><li class="c10 li-bullet-0"><span class="c7">Relative Standard Curve</span></li><li class="c10 li-bullet-0"><span class="c7">Comparative Ct (&Delta;&Delta;Ct)</span></li><li class="c10 li-bullet-0"><span class="c7">Melt Curve</span></li><li class="c10 li-bullet-0"><span class="c7">Genotyping</span></li><li class="c10 li-bullet-0"><span class="c7">Presence/Absence</span></li></ul><p class="c2"><span>The </span><span class="c3">container type</span><span>&nbsp;in all cases should be &ldquo;qPCR reaction block&rdquo;, while the </span><span class="c3">plate well count</span><span>&nbsp;should be taken from the </span><span class="c6">Block Type</span><span>&nbsp;field in the </span><span class="c6">header </span><span>block. The value of </span><span class="c3">plate well count</span><span class="c7">&nbsp;should be either 96 or 384.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">well volume</span><span class="c7">&nbsp;is not reported in the QuantStudio export, and can be ignored.</span></p><h3 class="c41" id="h.pkhvin69n6cv"><span class="c4">Measurement Document</span></h3><p class="c2"><span>Each </span><span class="c3">Measurement Document</span><span>&nbsp;in the would correspond to the </span><span class="c6">Target DNA</span><span>&nbsp;being amplified in the PCR reaction. qPCR reactions can be multiplexed with amplifications of multiple DNA targets within single reaction, with each represented as a separate </span><span class="c3">Measurement Document</span><span>&nbsp;within the </span><span class="c3">Measurement Aggregate Document</span><span>. In practice this multiplexing is typically limited to (2) DNA targets. Therefore, each </span><span class="c3">qPCR Document</span><span>&nbsp;will have either a single </span><span class="c3">Measurement Document</span><span>&nbsp;for single-plex assays or two </span><span class="c3">Measurement Documents</span><span class="c7">&nbsp;for multiplex assays.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">measurement identifier</span><span class="c7">&nbsp;is not derived from the instrument file and instead should be a universal unique identifier generated and added to the ASM file.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">measurement time</span><span>&nbsp;is recoded within the </span><span class="c6">Experiment Run End Time </span><span>field in the </span><span class="c6">header </span><span class="c7">block and should be recorded in a YYYY-MM-DD HH:MM:SS format.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">target DNA description</span><span>&nbsp;corresponds to the entry in the </span><span class="c6">Target Name </span><span>column within the </span><span class="c6">Sample Setup</span><span>&nbsp;block. In the case of a multiplexed experiment there will be multiple entries per </span><span class="c6">Well</span><span>&nbsp;in the </span><span class="c6">Sample Setup </span><span class="c7">block each with different DNA targets.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c22">Sample</span><span>&nbsp;</span><span class="c24 c34 c22">Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">Sample Document</span><span class="c7">&nbsp;contains details of the sample being assess as part of the measurement.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">sample identifier</span><span>, </span><span class="c3">sample role identifier</span><span>, and </span><span class="c3">well location identifier</span><span>, are recorded within the </span><span class="c6">Sample Setup </span><span>block, with the corresponding entries included under the </span><span class="c6">Sample Name</span><span>, </span><span class="c6">Task</span><span>, and </span><span class="c6">Well Position</span><span class="c7">&nbsp;columns.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">well plate identifier</span><span>&nbsp;if present will be recorded in the </span><span class="c6">Experiment Barcode</span><span>&nbsp;field in the </span><span class="c6">header </span><span class="c7">block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">batch identifier</span><span>&nbsp;and </span><span class="c3">mass concentration</span><span class="c7">&nbsp;properties are not reported in the QuantStudio export.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c24 c34 c22">Device Control Aggregate Document / Device Control Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">device control document</span><span class="c7">&nbsp;within the ASM is intended to record information about device settings under which each measurement was perform.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">device type</span><span class="c7">&nbsp;in all cases can be recorded as &ldquo;qPCR&rdquo;.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">device identifier</span><span>&nbsp;has been captured previously in the </span><span class="c3">Device System Document</span><span>&nbsp;and does not need to be recorded again here. The option for </span><span class="c3">device identifier</span><span class="c7">&nbsp;is provided at this level of the data model if one desires to capture the identify of a specific instrument component, that is a part of the larger device system, which is used to perform the measurement.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">measurement method identifier</span><span>&nbsp;is obtained from the </span><span class="c6">Quantification Cycle Method </span><span>within the </span><span class="c6">header </span><span class="c7">block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">total cycle number setting</span><span>&nbsp;is not reported directly, but this value can be determined by taking the largest number recorded within the </span><span class="c6">Cycle </span><span>column within the </span><span class="c6">Amplification Data</span><span>&nbsp;block. Note, that the </span><span class="c6">Amplification Data</span><span>&nbsp;block is indexed using </span><span class="c6">Well</span><span>&nbsp;rather than </span><span class="c6">Well Position</span><span>, where </span><span class="c6">Well </span><span>values of 1, 2, 3 correspond to </span><span class="c6">Well Position </span><span class="c7">values of A1, A2, A3.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>Besides </span><span class="c3">total cycle number setting</span><span>, the other PCR thermo-cycling parameters are not included in the QuantStudio output, so the entries for </span><span class="c3">denaturing</span><span>, </span><span class="c3">annealing</span><span>, and </span><span class="c3">extension</span><span class="c7">&nbsp;can be ignored.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The record for </span><span class="c3">qPCR Detection Chemistry</span><span>&nbsp;is provided in the </span><span class="c6">Chemistry</span><span>&nbsp;field within the </span><span class="c6">header </span><span class="c7">block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">reporter dye setting</span><span>&nbsp;and </span><span class="c3">quencher dye setting</span><span>&nbsp;are provided within the </span><span class="c6">Reporter</span><span>&nbsp;and </span><span class="c6">Quencher</span><span>&nbsp;columns within the </span><span class="c6">Sample Setup</span><span>&nbsp;block OR the </span><span class="c6">Results</span><span class="c7">&nbsp;block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">passive reference dye setting</span><span>&nbsp;is provided in the </span><span class="c6">Passive Reference</span><span>&nbsp;field within the </span><span class="c6">header </span><span class="c7">block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c24 c34 c22">Processed Data Aggregate Document / Processed Data Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">processed data document</span><span>&nbsp;contains results data which is derived from measurement data and who values are dependent on user defined parameters or transformation. The parameters which determine the results within the </span><span class="c3">processed data document</span><span>&nbsp;are recorded within the </span><span class="c3">data processing document</span><span class="c7">.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c24 c34 c22">Data Processing Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The entries for the </span><span class="c3">data processing document</span><span>&nbsp;are all found within the </span><span class="c6">Results</span><span class="c7">&nbsp;block table.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">automatic cycle threshold enable setting</span><span>, </span><span class="c3">cycle threshold value setting</span><span>, and </span><span class="c3">automatic baseline determination enabled setting</span><span>&nbsp;are provided under the </span><span class="c6">Automatic Ct Threshold, Ct Threshold, </span><span>and </span><span class="c6">Automatic Baseline </span><span class="c7">columns.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">baseline determination start cycle setting</span><span>&nbsp;and </span><span class="c3">baseline determination end cycle</span><span>&nbsp;can be excluded from the ASM. While there are </span><span class="c6">Baseline Start</span><span>&nbsp;and </span><span class="c6">Baseline End</span><span>&nbsp;columns within the </span><span class="c6">Results </span><span class="c7">block table, these do not reflect the setting values for these parameters in the originating QuantStudio experiment.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">genotyping determination method</span><span>, and </span><span class="c3">genotyping determination method setting</span><span>&nbsp;are only relevant in the context of </span><span class="c6">Presence/Absence</span><span>&nbsp;or </span><span class="c6">Genotyping</span><span>&nbsp;experiment types. Within </span><span class="c6">Presence/Absence </span><span>experiments the value within the </span><span class="c6">Threshold </span><span>column should be returned to the </span><span class="c3">genotyping determination method setting</span><span>, while for </span><span class="c6">Genotyping </span><span>experiments, the entry under the </span><span class="c6">Method </span><span>column should be returned to the </span><span class="c3">genotyping determination method</span><span class="c7">&nbsp;field.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c24 c22 c34">Remainder of Processed Data Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">cycle threshold result</span><span>, </span><span class="c3">normalized reported result</span><span>, </span><span class="c3">baseline corrected reported result</span><span>, and </span><span class="c3">genotyping determination result</span><span>&nbsp;are all reported as part of the </span><span class="c6">Results</span><span class="c7">&nbsp;block.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">cycle threshold result</span><span>, is recorded under the </span><span class="c6">CT </span><span>column and is expected to be present in any data set in which the </span><span class="c3">measurement method identifier</span><span>&nbsp;/ </span><span class="c6">Quantification Cycle Method </span><span class="c7">is &lsquo;Ct&rsquo;.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">normalized reporter result</span><span>&nbsp;is recorded under the </span><span class="c6">Rn</span><span>&nbsp;column while the </span><span class="c3">baseline corrected reported result</span><span>&nbsp;is recorded under the </span><span class="c6">Delta Rn </span><span>column. It is expected that one of these values be reported in </span><span class="c6">Presence/Absence</span><span class="c7">&nbsp;experiments.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">genotyping determination result</span><span>&nbsp;is recorded under the </span><span class="c6">Call </span><span>column, and is expected to be reported in </span><span class="c6">Presence/Absence</span><span>&nbsp;and </span><span class="c6">Genotyping </span><span class="c7">experiments.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The data for the </span><span class="c3">normalized reporter data cube</span><span>&nbsp;and </span><span class="c3">baseline corrected data cube</span><span>&nbsp;are recorded within the </span><span class="c6">Amplification Data</span><span>&nbsp;block, with the </span><span class="c3">normalized reporter</span><span>&nbsp;data under the </span><span class="c6">Rn </span><span>column and the </span><span class="c3">baseline corrected</span><span>&nbsp;data under the </span><span class="c6">Delta Rn</span><span class="c7">&nbsp;column.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c24 c34 c22">Remainder of Measurement Document</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The remainder of the </span><span class="c3">Measurement Document</span><span>&nbsp;contains the </span><span class="c3">reporter dye data cube</span><span>, </span><span class="c3">passive reference dye data cube</span><span>, and </span><span class="c3">melt curve data cube</span><span class="c7">.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The data for </span><span class="c3">reporter dye data cube</span><span>&nbsp;and </span><span class="c3">passive reference dye data cube</span><span>&nbsp;are recorded within the </span><span class="c6">Multicomponent</span><span>&nbsp;block, with the column headers corresponding to the </span><span class="c3">reporter dye setting</span><span>&nbsp;and the </span><span class="c3">passive reference dye setting</span><span class="c7">&nbsp;respectively.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The data for the </span><span class="c3">melt curve data cube</span><span>&nbsp;are recorded within the </span><span class="c6">Melt Curve Raw Data </span><span>block. Note, that similar to the </span><span class="c6">Amplification Data</span><span>&nbsp;block, the </span><span class="c6">Melt Curve raw Data</span><span>&nbsp;block is indexed using </span><span class="c6">Well</span><span>&nbsp;rather than </span><span class="c6">Well Position</span><span>, where </span><span class="c6">Well </span><span>values of 1, 2, 3 correspond to </span><span class="c6">Well Position </span><span class="c7">values of A1, A2, A3.</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.q6gxnwi5qqtv"><span class="c4">Formatting of &lsquo;Genotyping&rsquo; experiments</span></h3><p class="c2"><span>The Experiment Type </span><span class="c6">Genotyping</span><span>&nbsp;is variant of the multiplexing assays performed by QuantStudio. The file exports from </span><span class="c6">Genotyping </span><span>experiments have an alternative formatting of the </span><span class="c6">Results</span><span>&nbsp;block table compared other Experiment Types. The </span><span class="c6">Genotyping</span><span>&nbsp;experiments have (2) </span><span class="c3">Target DNA </span><span>per well, similar to other multiplexing experiments, however, the data for each </span><span class="c3">Target DNA</span><span>&nbsp;is reported on the same row, rather than on separate rows as is done with other multiplexing experiments. To conform to the pattern used in the ASM for all other experiment types, these fields need to be segmented to separate </span><span class="c3">measurement documents</span><span>&nbsp;(within the same </span><span class="c3">qPCR document</span><span class="c7">).</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>With the </span><span class="c6">Genotyping</span><span>&nbsp;experiments, the two </span><span class="c3">Target DNA</span><span>&nbsp;are referred to as </span><span class="c6">Allele 1</span><span>&nbsp;and </span><span class="c6">Allele 2</span><span>. The table below details the columns within the </span><span class="c6">Results</span><span>&nbsp;block table which have the </span><span class="c3">Target DNA</span><span class="c7">&nbsp;appended.</span></p><p class="c0"><span class="c7"></span></p><a id="t.abf11f0faeb4664adc7c10c01f3fda19374e15d2"></a><a id="t.0"></a><table class="c25"><tr class="c26"><td class="c37" colspan="1" rowspan="1"><p class="c27"><span class="c22">Results Block Column</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c27"><span class="c22">Column Base Name</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c27"><span class="c22">ASM Field</span></p></td></tr><tr class="c31"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Delta Rn / Allele 2 Delta Rn</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Rn</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">baseline corrected reporter result</span></p></td></tr><tr class="c42"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Automatic Ct Threshold / Allele 2 Automatic Ct Threshold</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Automatic Ct Threshold</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">automatic cycle threshold enabled setting</span></p></td></tr><tr class="c31"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Ct Threshold /</span></p><p class="c2"><span class="c7">Allele 2 Ct Threshold</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct Threshold</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold value setting</span></p></td></tr><tr class="c31"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Baseline Start /</span></p><p class="c2"><span class="c7">Allele 2 Baseline Start</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Baseline Start</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">basline determination start cycle</span></p></td></tr><tr class="c31"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Baseline End /</span></p><p class="c2"><span class="c7">Allele 2 Baseline End</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Baseline End</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">basline determination end cycle</span></p></td></tr><tr class="c26"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Ct / Allele 2 Ct</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td></tr><tr class="c31"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Amp Score /</span></p><p class="c2"><span class="c7">Allele 2 Amp Score</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Amp Score</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">N/A</span></p></td></tr><tr class="c31"><td class="c37" colspan="1" rowspan="1"><p class="c2"><span class="c7">Allele 1 Cq Confidence /</span></p><p class="c2"><span class="c7">Allele 2 Cq Confidence</span></p></td><td class="c36" colspan="1" rowspan="1"><p class="c2"><span class="c7">Cq Confidence</span></p></td><td class="c14" colspan="1" rowspan="1"><p class="c2"><span class="c7">N/A</span></p></td></tr></table><p class="c2"><span>The individual values for </span><span class="c6">Delta Rn</span><span>, </span><span class="c6">Automatic Ct Threshold</span><span>, </span><span class="c6">Ct Threshold</span><span>, </span><span class="c6">Baseline Start</span><span>, </span><span class="c6">Baseline End, </span><span>and </span><span class="c6">Ct</span><span>&nbsp;should be placed into the separate </span><span class="c3">measurement documents</span><span>. While </span><span class="c6">Amp Score </span><span>and </span><span class="c6">Cq Confidence</span><span class="c7">&nbsp;are not currently recored within the ASM and can be ignored.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>Within these separate </span><span class="c3">measurement documents</span><span>&nbsp;the entry for </span><span class="c3">Target DNA description</span><span>&nbsp;should a concatenation of the entry under </span><span class="c6">SNP Assay Name</span><span>&nbsp;and the </span><span class="c6">Allele, </span><span>as shown in the example: CY19_2 Allele 1. Both </span><span class="c3">measurement documents</span><span>&nbsp;would included the same </span><span class="c3">sample identifier</span><span class="c7">&nbsp;since that is common to both measurements.</span></p><p class="c0"><span class="c7"></span></p><h3 class="c41" id="h.h3p9fke4np3g"><span class="c4">Calculated Data Aggregate / Calculated Data Document</span></h3><p class="c2"><span>Within the ASM, the </span><span class="c3">calculated data document</span><span>&nbsp;is used to record results that are derived from calculations performed using data </span><span class="c6">across</span><span>&nbsp;different </span><span class="c3">measurement documents</span><span>&nbsp;and/or </span><span class="c3">other calculated data documents</span><span>&nbsp;. A </span><span class="c3">calculated data document</span><span>&nbsp;differs from a </span><span class="c3">processed data document</span><span>&nbsp;in that a </span><span class="c3">processed data document</span><span>&nbsp;is a facet of a </span><span class="c3">measurement document</span><span>&nbsp;and contains results derived from transformation of the data </span><span class="c6">within</span><span>&nbsp;that </span><span class="c3">measurement document</span><span class="c7">.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>While a </span><span class="c3">calculated data document</span><span>&nbsp;records calculated results, </span><span class="c6 c22">the adapter is not expected to perform any calculations</span><span class="c7">, and rather parse calculated values which appear within the QuantStudio file export.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">calculated data document</span><span>&nbsp;includes a </span><span class="c3">calculated data identifier</span><span>, which similar to the </span><span class="c3">measurement identifier</span><span class="c7">&nbsp;is not derived from the instrument file and instead should be a universal unique identifier generated and added to the ASM file.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">calculated data document</span><span>&nbsp;also includes a </span><span class="c3">data source aggregate document/data source document</span><span>&nbsp;which details the inputs to the calculation including: a </span><span class="c3">data source identifier</span><span>&nbsp;used to reference another </span><span class="c3">measurement identifier</span><span>&nbsp;or </span><span class="c3">calculated data identifier</span><span>, and a </span><span class="c3">data source feature</span><span class="c7">&nbsp;which is used to reference the field within the source document which was used as input to the calculation.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">calculated data document</span><span>&nbsp;may contain a </span><span class="c3">data processing document</span><span>&nbsp;which will records needed to interpret how the calculation was performed. For the qPCR ASM, this </span><span class="c3">data processing document</span><span>&nbsp;could include </span><span class="c3">reference DNA description</span><span>&nbsp;and </span><span class="c3">reference sample description</span><span>&nbsp;which are needed to interpret </span><span class="c6">Delta Ct</span><span>&nbsp;and </span><span class="c6">Delta Delta Ct</span><span class="c7">&nbsp;calculations (discussed below).</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">calculated data document</span><span>&nbsp;also includes a </span><span class="c3">calculated data name</span><span>, a name for the calculated result; optionally a </span><span class="c3">calculated data description</span><span>&nbsp;to record how the calculation was performed; and a </span><span class="c3">calculated datum</span><span class="c7">&nbsp;to record the calculated result.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The table below contains the fields within the QuantStudio export that should be captured within </span><span class="c3">calculated data documents</span><span>. These fields, when present, will appear within the </span><span class="c6">Results</span><span class="c7">&nbsp;block in the output file.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The ability of a </span><span class="c3">calculated data document</span><span>&nbsp;to reference other </span><span class="c3">calculated data identifiers</span><span>&nbsp;and/or </span><span class="c3">measurement identifiers</span><span>&nbsp;allows for the capturing the relationships between calculated data and the original measurements. The relationships of the calculated field in the QuantStudio export are detailed in the </span><span class="c6">Data Source Feature</span><span>&nbsp;and </span><span class="c6">Aggregating Property</span><span class="c7">&nbsp;columns in the table.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>Note: that the </span><span class="c6">Aggregating Property</span><span class="c7">&nbsp;is not explicitly reported in the QuantStudio export file, but rather is derived from documentation about the calculations performed.</span></p><a id="t.85073a2ee17a92f62952064d0e5258ed8e837355"></a><a id="t.1"></a><table class="c54"><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c22">QuantStudio Column</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c22">Calculated Data Name</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c22">Data Source</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c22">Data Source ID</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c22">Data Source Feature</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c22">Data Processing</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c22">Aggregating Property</span></p></td></tr><tr class="c48"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result, Y-intercept, Slope</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td></tr><tr class="c31"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity Mean</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity Mean</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA</span></p></td></tr><tr class="c31"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity SD</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity SD</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct Mean</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct Mean</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct SD</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct SD</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct Mean</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct Mean</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct Mean</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">reference DNA description</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA*</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct SE</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct SE</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct Mean</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">reference DNA description</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA*</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Delta Ct</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Delta Ct</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct Mean</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">reference sample description</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA*</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ</span></p><p class="c2"><span class="c7">(comparative Ct)</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Delta Ct</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA*</span></p></td></tr><tr class="c49"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ</span></p><p class="c2"><span class="c7">(relative standard curve)</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity Mean</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">reference DNA description, reference sample description</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sampe id, target DNA*</span></p></td></tr><tr class="c31"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ min</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ min</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA*</span></p></td></tr><tr class="c31"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ max</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ max</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">calculated data id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA*</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rn Mean</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rn Mean</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">normalized reporter result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA</span></p></td></tr><tr class="c42"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rn SD</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rn SD</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">normalized reporter result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">sample id, target DNA</span></p></td></tr><tr class="c17"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Y-Intercept</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Y-Intercept</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">target DNA, well role type == STANDARD</span></p></td></tr><tr class="c17"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">R(superscript 2)</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rsquared</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">target DNA, well role type == STANDARD</span></p></td></tr><tr class="c17"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Slope</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Slope</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">target DNA, well role type == STANDARD</span></p></td></tr><tr class="c17"><td class="c29" colspan="1" rowspan="1"><p class="c2"><span class="c7">Efficiency</span></p></td><td class="c18" colspan="1" rowspan="1"><p class="c2"><span class="c7">Efficiency</span></p></td><td class="c12" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement doc</span></p></td><td class="c35" colspan="1" rowspan="1"><p class="c2"><span class="c7">measurement id</span></p></td><td class="c9" colspan="1" rowspan="1"><p class="c2"><span class="c7">cycle threshold result</span></p></td><td class="c19" colspan="1" rowspan="1"><p class="c2"><span class="c7">None</span></p></td><td class="c16" colspan="1" rowspan="1"><p class="c2"><span class="c7">target DNA, well role type == STANDARD</span></p></td></tr></table><p class="c2"><span class="c24 c34 c22">Relationship between calculated values:</span></p><p class="c2"><span class="c7">cycle threshold result (measurement doc) &rarr; Ct Mean &rarr; Delta Ct Mean &rarr; Delta Delta Ct &rarr; RQ</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c22">Sample Calculation of </span><span class="c6 c22">Delta Delta Ct </span><span class="c22">and </span><span class="c6 c22 c24">RQ</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>Delta Ct = Ct Mean - Ct Mean </span><span class="c34 c33 c56">(ref target)</span></p><p class="c2"><span class="c7">Delta Delta Ct = Delta Ct - Delta Ct (ref sample)</span></p><p class="c2"><span>RQ = 2^ </span><span class="c34 c33 c53">-(Delta Delta Ct)</span></p><a id="t.76bceb84d03793a877dde375480c1c007d4b52c1"></a><a id="t.2"></a><table class="c25"><tr class="c31"><td class="c39" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c30" colspan="1" rowspan="1"><p class="c2"><span class="c22">DNA Target A (Ct Mean)</span></p></td><td class="c45" colspan="1" rowspan="1"><p class="c2"><span class="c22">DNA Target B (Ct Mean)</span></p></td><td class="c13" colspan="1" rowspan="1"><p class="c2"><span class="c22">Delta Ct</span></p></td><td class="c43" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td></tr><tr class="c26"><td class="c39" colspan="1" rowspan="1"><p class="c15"><span class="c22">Sample 1</span></p></td><td class="c30" colspan="1" rowspan="1"><p class="c15"><span class="c7">27</span></p></td><td class="c45" colspan="1" rowspan="1"><p class="c15"><span class="c7">25</span></p></td><td class="c13" colspan="1" rowspan="1"><p class="c15"><span class="c7">-2</span></p></td><td class="c43" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td></tr><tr class="c26"><td class="c39" colspan="1" rowspan="1"><p class="c15"><span class="c22">Sample 2</span></p></td><td class="c30" colspan="1" rowspan="1"><p class="c15"><span class="c7">27</span></p></td><td class="c45" colspan="1" rowspan="1"><p class="c15"><span class="c7">26</span></p></td><td class="c13" colspan="1" rowspan="1"><p class="c15"><span class="c7">-1</span></p></td><td class="c43" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td></tr><tr class="c38"><td class="c39" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c30" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c45" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c13" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c43" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td></tr><tr class="c26"><td class="c39" colspan="1" rowspan="1"><p class="c15"><span class="c22">Reference Target</span></p></td><td class="c30" colspan="1" rowspan="1"><p class="c2"><span class="c7">Target A</span></p></td><td class="c45" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c13" colspan="1" rowspan="1"><p class="c15"><span class="c7">1</span></p></td><td class="c43" colspan="1" rowspan="1"><p class="c2"><span class="c22">Delta Delta Ct</span></p></td></tr><tr class="c31"><td class="c39" colspan="1" rowspan="1"><p class="c15"><span class="c22">Reference Sample</span></p></td><td class="c30" colspan="1" rowspan="1"><p class="c2"><span class="c7">Sample 1</span></p></td><td class="c45" colspan="1" rowspan="1"><p class="c0"><span class="c7"></span></p></td><td class="c13" colspan="1" rowspan="1"><p class="c15"><span class="c7">0.5</span></p></td><td class="c43" colspan="1" rowspan="1"><p class="c2"><span class="c22">RQ</span></p></td></tr></table><p class="c2"><span>As shown in</span><span><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1xVlLbxejUSnITq_BjI9t25SFDblXXyTr/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185304557&amp;usg=AOvVaw0Im5mdZew7lPgWk1htgwZE">&nbsp;</a></span><span class="c5"><a class="c32" href="https://www.google.com/url?q=https://drive.google.com/file/d/1xVlLbxejUSnITq_BjI9t25SFDblXXyTr/view?usp%3Ddrive_link&amp;sa=D&amp;source=editors&amp;ust=1715829185304890&amp;usg=AOvVaw092678eYC20OiOCBPhClO9">Thermo_QuantStudio_example01_asm_rev03.json</a></span><span>&nbsp;the </span><span class="c3">calculated data document</span><span>&nbsp;for </span><span class="c6">Delta Delta Ct</span><span class="c7">&nbsp;is derived from the following:</span></p><ol class="c11 lst-kix_i97szj75x3di-0 start" start="1"><li class="c10 li-bullet-0"><span>A </span><span class="c3">calculated data document</span><span>&nbsp;for </span><span class="c6">Ct Mean</span><span>&nbsp;which references as </span><span class="c3">data sources</span><span>&nbsp;the </span><span class="c3">cycle threshold results</span><span>&nbsp;from all </span><span class="c3">measurement documents</span><span>&nbsp;sharing the same </span><span class="c3">sample identifier</span><span>&nbsp;and </span><span class="c3">target DNA description</span><span class="c7">.</span></li></ol><p class="c2"><span class="c6">Note:</span><span>&nbsp;In this example file there are no replicates of </span><span class="c3">sample identifier </span><span>and </span><span class="c3">target DNA</span><span>, so the there is only a single </span><span class="c3">data source document</span><span>&nbsp;and the value of </span><span class="c6">Ct Mean </span><span>equals the </span><span class="c3">cycle threshold result</span><span>&nbsp;since this reflects a </span><span class="c6">Mean </span><span>of </span><span class="c24 c33 c6">n=1</span></p><ol class="c11 lst-kix_khhpt4pbx05-0 start" start="1"><li class="c10 li-bullet-0"><span>A </span><span class="c3">calculated data document</span><span>&nbsp;for </span><span class="c6">Delta Ct Mean</span><span>&nbsp;which references as </span><span class="c3">data sources</span><span>&nbsp;the two </span><span class="c3">calculated data documents</span><span>&nbsp;for </span><span class="c6">Ct Mean</span><span>&nbsp;from (#1) which reference </span><span class="c3">measurement documents</span><span>&nbsp;with the same </span><span class="c1">sample identifier</span></li><li class="c10 li-bullet-0"><span>A </span><span class="c3">calculated data document</span><span>&nbsp;for </span><span class="c6">Delta Delta Ct</span><span>&nbsp;which references as </span><span class="c3">data sources</span><span>&nbsp;the </span><span class="c3">calculated data document</span><span>&nbsp;for </span><span class="c6">Delta Ct Mean</span><span>&nbsp;from (#2) and the </span><span class="c3">calculated data document</span><span>&nbsp;for </span><span class="c6">Delta Ct Mean </span><span>- also from #2 - but </span><span class="c6">specifically</span><span>&nbsp;for the </span><span class="c3">sample identifier</span><span>&nbsp;that corresponds to the </span><span class="c1">reference sample description</span></li></ol><p class="c0"><span class="c7"></span></p><p class="c2"><span>The records for </span><span class="c3">reference DNA description</span><span>&nbsp;and </span><span class="c3">reference sample description</span><span>&nbsp;appear at the bottom of the </span><span class="c6">Results </span><span>block. The </span><span class="c3">reference DNA description</span><span>&nbsp;appears in the field </span><span class="c6">Endogenous Control</span><span>&nbsp;and </span><span class="c3">reference sample description</span><span>&nbsp;in the field </span><span class="c6">Reference Sample</span><span class="c7">.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>For the calculated fields: </span><span class="c6">Delta Ct Mean, Delta Ct SE, Delta Delta Ct, RQ, RQ min, RQ max</span><span>&nbsp;there will be no value reported for those rows where the </span><span class="c3">Target DNA </span><span>equals the </span><span class="c6">Endogenous Control / </span><span class="c3">reference DNA description</span><span class="c7">&nbsp;.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The </span><span class="c3">calculated data documents</span><span>&nbsp;for other calculated fields including </span><span class="c6">Y-Intercept, R2, Slope, </span><span>and </span><span class="c6">Efficiency</span><span>&nbsp;will be reference as </span><span class="c3">data sources</span><span>&nbsp;the </span><span class="c3">measurement documents</span><span>&nbsp;that share the same </span><span class="c3">target DNA</span><span>&nbsp;and have the </span><span class="c3">well role type</span><span class="c7">&nbsp;of STANDARD.</span></p><p class="c2"><span class="c7">The</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span>The table below details which calculated values are expected to be present in the </span><span class="c6">Results</span><span>&nbsp;block table of different </span><span class="c6">Experiment Types. </span><span class="c7">The QuantStudio UI allows user to select / deselect columns for export, so even though a particular calculated value may be generated from a certain experiment types, exports of that experiment type many not always contain that particular calculated value.</span></p><p class="c0"><span class="c7"></span></p><a id="t.220272f5f938e207ff66f91baebe7a915a374ac1"></a><a id="t.3"></a><table class="c25"><tr class="c31"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c22">QuantStudio Column</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c22">Experiment Type</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity Mean</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Quantity SD</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct Mean</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Ct SD</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct Mean</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Ct SE</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Delta Delta Ct</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ*</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ min</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">RQ max</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Comparative Ct, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rn Mean</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Presence/Absence</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Rn SD</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Presence/Absence</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Y-Intercept</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">R(superscript 2)</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Slope</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Standard Curve, Relative Standard Curve</span></p></td></tr><tr class="c26"><td class="c21" colspan="1" rowspan="1"><p class="c2"><span class="c7">Efficiency</span></p></td><td class="c20" colspan="1" rowspan="1"><p class="c2"><span class="c7">Standard Curve, Relative Standard Curve</span></p></td></tr></table><p class="c2"><span class="c7">Note that the calculation of RQ differs between Comparative Ct and Relative Standard Curve experiments.</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">For Comparative Ct the relationship of calculations is:</span></p><p class="c2"><span class="c7">cycle threshold result (measurement doc) &rarr; Ct Mean &rarr; Delta Ct Mean &rarr; Delta Delta Ct &rarr; RQ</span></p><p class="c0"><span class="c7"></span></p><p class="c2"><span class="c7">For Relative Standard Curve the relationship of calculations is:</span></p><p class="c2"><span class="c7">cycle threshold result (measurement doc) &rarr; Quantity &rarr; Quantity Mean &rarr; RQ</span></p><h3 class="c41" id="h.hm2tl9pwlk7b"><span class="c4">Change Log:</span></h3><p class="c2"><span class="c24 c33 c6">2023-07-31 Rev 01:</span></p><p class="c0"><span class="c7"></span></p><ul class="c11 lst-kix_xwb2xgsmkc7s-0 start"><li class="c10 li-bullet-0"><span class="c7">Initial version</span></li></ul><p class="c2"><span class="c24 c33 c6">2023-08-22 Rev 02:</span></p><p class="c0"><span class="c7"></span></p><ul class="c11 lst-kix_rysblip21ldc-0 start"><li class="c10 li-bullet-0"><span class="c7">Updated with second instrument output/ASM example.</span></li></ul><p class="c2"><span class="c24 c33 c6">2023-08-29 Rev 03:</span></p><ul class="c11 lst-kix_e63skeaub1i7-0 start"><li class="c10 li-bullet-0"><span>Updated document to reflect the use of </span><span class="c3">Calculated Data Document</span><span class="c7">&nbsp;within the ASM</span></li><li class="c10 li-bullet-0"><span class="c7">Included revised version of qPCR ASM diagram, simplified schema, and example 01 output.</span></li><li class="c10 li-bullet-0"><span>Updated ASM diagram and schema to reference </span><span class="c6">well plate identifier</span><span>&nbsp;instead of </span><span class="c6">plate barcode</span><span class="c7">.</span></li><li class="c10 li-bullet-0"><span>Included description of how to processing </span><span class="c6">Experiment Type - Genotyping</span><span class="c7">&nbsp;exports with accompanying example.</span></li></ul><p class="c2"><span class="c24 c33 c6">2023-09-28 Rev 04:</span></p><ul class="c11 lst-kix_ova5lqw9xm1i-0 start"><li class="c10 li-bullet-0"><span>Updated descriptions of </span><span class="c3">Calculated Data Document</span><span class="c7">&nbsp;to account for different methods of calculation of RQ between Comparative Ct and Relative Standard Curve experiments.</span></li></ul><p class="c2"><span class="c24 c33 c6">2024-01-19 Rev 05:</span></p><ul class="c11 lst-kix_2u1s0aut9bz5-0 start"><li class="c10 li-bullet-0"><span>Removed references to use of </span><span class="c3">calculated data aggregate / calculated data documents</span><span>&nbsp;within </span><span class="c3">qPCR document</span><span>, to reflect practice of </span><span class="c3">calculated data aggregate / calculated data documents</span><span>&nbsp;only being implemented at the </span><span class="c6">technique aggregate </span><span>level (i.e. </span><span class="c3">qpcr aggregate document</span><span class="c7">)</span></li></ul><p class="c0"><span class="c7"></span></p></body></html>